File: 23.合并-k-个升序链表.cpp
/*
 * @lc app=leetcode.cn id=23 lang=cpp
 *
 * [23] 合并 K 个升序链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
// #include <vector>
// #include <queue>
// struct ListNode {
//     int val;
//     ListNode *next;
//     ListNode() : val(0), next(nullptr) {}
//     ListNode(int x) : val(x), next(nullptr) {}
//     ListNode(int x, ListNode *next) : val(x), next(next) {}
// };
class Solution {
public:
    ListNode* mergeKLists(std::vector<ListNode*>& lists) {
        if(lists.empty()) return nullptr;
        auto cmp = [](ListNode* l1, ListNode* l2) {
            return l1->val > l2->val;
        };
        std::priority_queue<ListNode*, std::vector<ListNode*>, decltype(cmp)> pq(cmp);
        for(auto l : lists){
            if(l != nullptr) pq.push(l);
        }
        ListNode* dummy = new ListNode(-1);
        ListNode* cur = dummy;
        while(!pq.empty()){
            auto node = pq.top();
            pq.pop();
            cur->next = node;
            cur = cur->next;
            if(node->next != nullptr){
                pq.push(node->next);
                node->next = nullptr;
            }
        }
        return dummy->next;
    }
};
// @lc code=end

File: 23.合并-k-个升序链表.java
/*
 * @lc app=leetcode.cn id=23 lang=java
 *
 * [23] 合并 K 个升序链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
import java.util.*;

class Solution {
    public static class ListNode{
        int val;
        ListNode next;
        ListNode(){}
        ListNode(int x){
            this.val = x;
        }
        ListNode(int x, ListNode next){
            this.val = x;
            this.next = next;
        }
    }

    public static class Node{
        int val;
        List<Node> children;
        Node(){
            // this.children = new ArrayList<>();
        }
        Node(int x){
            this.val = x;
            this.children = new ArrayList<>();
        }
        public void addNode(Node child){
            this.children.add(child);
        }
    }

    public ListNode mergeKLists(ListNode[] lists) {
        Comparator<ListNode> cmp = new Comparator<ListNode>() {
            @Override
            public int compare(ListNode p1, ListNode p2){
                return p1.val - p2.val;
            }
        };
        PriorityQueue<ListNode> myQueue = new PriorityQueue<>(cmp);
        int listCount = lists.length;
        for(int i = 0; i < listCount; i++){
            if(lists[i] != null){
                myQueue.add(lists[i]);
            }
        }
        ListNode dummyHead = new ListNode(-1);
        ListNode cur = dummyHead;
        while(!myQueue.isEmpty()){
            ListNode temp = myQueue.poll();
            cur.next = temp;
            cur = cur.next;
            if(temp.next != null){
                myQueue.add(temp.next);
            }
            temp.next = null;
        }
        return dummyHead.next;
    }
}
// @lc code=end

